test/extensions/filters/http/cache/http_cache_implementation_test_common.cc

```diff
@@ -0,0 +1,749 @@
+}
+
+void HttpCacheImplementationTest::updateHeaders(
+    absl::string_view request_path, const Http::TestResponseHeaderMapImpl& response_headers,
+    const ResponseMetadata& metadata) {
+  LookupContextPtr lookup_context = lookup(request_path);
+  cache().updateHeaders(*lookup_context, response_headers, metadata);
+}
+
+LookupContextPtr HttpCacheImplementationTest::lookup(absl::string_view request_path) {
+  LookupRequest request = makeLookupRequest(request_path);
+  LookupContextPtr context = cache().makeLookupContext(std::move(request), decoder_callbacks_);
+  absl::Notification lookup_result_received;
+  context->getHeaders([this, &lookup_result_received](LookupResult&& result) {
+    lookup_result_ = std::move(result);
+    lookup_result_received.Notify();
+  });
+  EXPECT_TRUE(lookup_result_received.WaitForNotificationWithTimeout(absl::Seconds(3)));
+  return context;
+}
+
+absl::Status HttpCacheImplementationTest::insert(LookupContextPtr lookup,
+                                                 const Http::TestResponseHeaderMapImpl& headers,
+                                                 const absl::string_view body,
+                                                 std::chrono::milliseconds timeout) {
+  return insert(std::move(lookup), headers, body, absl::nullopt, timeout);
+}
+
+absl::Status HttpCacheImplementationTest::insert(
+    LookupContextPtr lookup, const Http::TestResponseHeaderMapImpl& headers,
+    const absl::string_view body, const absl::optional<Http::TestResponseTrailerMapImpl> trailers,
+    std::chrono::milliseconds timeout) {
+  InsertContextPtr inserter = cache().makeInsertContext(std::move(lookup), encoder_callbacks_);
+  const ResponseMetadata metadata{time_system_.systemTime()};
+  bool headers_end_stream = body.empty() && !trailers.has_value();
+  inserter->insertHeaders(headers, metadata, headers_end_stream);
+
+  if (headers_end_stream) {
+    return absl::OkStatus();
+  }
+  if (!body.empty() && !trailers.has_value()) {
+    inserter->insertBody(Buffer::OwnedImpl(body), nullptr,
+                         /*end_stream=*/!trailers.has_value());
+    return absl::OkStatus();
+  }
+  if (body.empty() && trailers.has_value()) {
+    inserter->insertTrailers(trailers.value());
+    return absl::OkStatus();
+  }
+
+  // For responses with body and trailers, wait for insertBody's callback before
+  // calling insertTrailers.
+  auto insert_mutex = std::make_shared<absl::Mutex>();
+  bool insert_done = false;
+  auto insert_cancelled = std::make_shared<bool>(false);
+  absl::Status insert_status;
+  InsertCallback insert_callback = [&trailers, &inserter, insert_mutex, &insert_done,
+                                    insert_cancelled, &insert_status](bool success_ready_for_more) {
+    absl::MutexLock lock(insert_mutex.get());
+    if (*insert_cancelled) {
+      return;
+    }
+    if (!success_ready_for_more) {
+      insert_status = absl::UnknownError("Insert was aborted by cache");
+      insert_done = true;
```

Comment:
what happens if there's a timeout after the insert_cancelled check above, and right before this line? Won't this write to an invalid stack address?

WDYT about bundling the insert mutex/done/cancelled bits into a structure and keeping a shared__ptr to that?

Reply:
We're holding insert_mutex for the duration of this callback, so there can't be a timeout in the middle.